codalab/model/bundle_model.py

```diff
@@ -677,6 +677,85 @@ def get_bundle_uuids(self, conditions, max_results):
 
         return self._execute_query(query)
 
+    def get_memoized_bundles(self, user_id, command, dependencies):
+        '''
+        Get the most recently finished bundle that matches with input command and dependencies.
+        :param user_id: a string that specifies the current user id.
+        :param command: a string that defines the command that is used to search for memoized bundles in the database.
+        :param dependencies: a list of string in the structure of ["key1:uuid1","key2:uuid2"] defines
+                            dependencies that are used to search for memoized bundles in the database.
+        :return: a list of matched uuids.
+        '''
+        # Get a list of uuids that matches with the given dependencies from the cl_bundle table.
+        # This step is to standardize uuid for future comparison.
+        with self.engine.begin() as connection:
+            # Select bundles that has the given command from the bundle table.
+            filter_on_command = (
+                select([cl_bundle.c.uuid])
+                .distinct()
+                .select_from(cl_bundle)
+                .where(and_(cl_bundle.c.command == command, cl_bundle.c.owner_id == user_id))
+                # Ensure the order of the returning bundles will be in the order of they were created.
+                .order_by(cl_bundle.c.id)
+            )
+            rows = connection.execute(filter_on_command).fetchall()
+            if not rows:
+                return []
+            uuids = [row[0] for row in rows]
+            # When there is no dependency to be matched, the target memozied bundle
```

Comment:
Is there a way to simplify the logic below?

Reply:
Did more tests and currently don't think we can combine those two conditions (with dependencies and without dependencies) into one together due to the following reason:
1. When there is no dependency specified, we want to search for the bundles that ONLY exist in the `bundle` table, but NOT in the `bundle_dependency` table.
1. When there are dependencies specified, we want to do an **inner join** between the `bundle` table and the `bundle_dependency` table, so that we will be able to select those bundles that match with input command as well as input dependencies.

Essentially the above two conditions are different when reflected on SQL query, but I may not see the optimal solution, let me know if I miss anything here.

Reply:
@percyliang Based on our discussion yesterday, I altered the database query logic to use join query on all dependencies. Changes are listed at https://github.com/codalab/codalab-worksheets/pull/2148. 
For the use case when there is no dependency provided, if we do want to include functionality to handle this, I don't think the logic can be combined with the one with dependencies. On the other hand, if we really don't want to handle this case, then I'll remove this piece of logic. 
I personally think providing the ability to return a memoized bundle without dependencies would be helpful especially in the case when the bundle might take a long time to finish. Suppose one user submitted a bundle that took a few hours to finish, after 1 month or so, he/she wanted to run it again, then they could really take advantage of using the memozie option. Let me know how you think!

Reply:
At the same time, I am benchmarking those queries in staging using a production size database just to make sure the performance of queries is good.

Reply:
@percyliang I later realized that [union on joins](#2148) could be avoided and use 1 JOIN query a time. The reason is that the list of dependencies that passed into `get_memoized_bundle` is already deduped. As all dependencies are unique now, when we do GROUP BY `child_uuid`, we just need to aggregate the number of records per `child_uuid` and then compare it with the number of dependencies. If the numbers match, then we find a memoized bundle.